### DISCUSSION TITLE
Adding RTC?

### DISCUSSION THREAD

**Discussion by majorpbx** [sanni/cartreader#255] [Ideas & Suggestions]
I'm not sure how feasible this is but can an RTC be added so that file times aren't all the same? Plus it will help across the board with remembering when you dumped which copy of what game.

Here is my work up of the board. It combines 2 Adafruit boards (Clock Gen Si5351A and RTC DS3231) into one and removes unnecessary solder pads. Looking into changing the DS3231 to DS1307 as the chip is cheaper. Essentially, both boards can be combined as they use different addresses. I'm going to bread board this and see if the Arduino sees all 3 SDA/SCL addresses from the OLED, Clock Gen, and RTC.

Board Dimensions: 25 mm x 34 mm

Front
![Board_Front](https://user-images.githubusercontent.com/18235905/133906439-f6588800-c15d-4074-8c1a-ef0e076d7c46.png)

Back
![Board_Back](https://user-images.githubusercontent.com/18235905/133906444-c4084d26-d513-44d5-a337-2a88ebb35d51.png)

**Comment by Dakkaron** [comment]
I like the idea, but wouldn't it be cheaper to just add a ready-made module instead of creating a new one? It would cut down on soldering work and mlst likely also cost.

**Reply by majorpbx** [reply]
It would and definitely preferable. But given space requirements of this project and availability of ready made modules just for this purpose is the reason i decided to tackle it.

**Reply by Dakkaron** [reply]
In that case, might it be better to directly integrate the rtc components onto the original board?

**Reply by majorpbx** [reply]
that was my alternate thought as well as throwing in the Clock Gen. The only thing the Clock Gen board does beyond the obvious is power regulate to 3V power and 3V for SCL/SDA too.

The only thing to consider is DS1307 is 5V and DS3231 is 3V. Though datasheets for both suggest they operate at 3-5V.

**Reply by majorpbx** [reply]
Technically speaking if the Sanni board conditions the power which it does if you have the battery add-on all you need is these chips added to the main board. Extra power conditioning for the SCL/SDA might be better in the long run but possibly omitted if the need to save space on the board exists.

DS3231 or DS1307 (not as accurate but could save space on the board unless you use a smaller DS3231)
and
Si5351 with 25MHz crystal

**Reply by majorpbx** [reply]
This leads me to my question of whether the board in its current state can be programmed to auto detect the cart that exists. My thought is if more than one exist can it prompt you to choose which to dump. But that will likely require a few more parts on the PCB to allow for choice control of the data lines. Just a wild thought!

**Reply by majorpbx** [reply]
Well looks like my design at least works! :grin: 

Scanning...
I2C device found at address 0x3C  !
I2C device found at address 0x60  !
I2C device found at address 0x68  !
done

**Reply by majorpbx** [reply]
Tested with the assistance of this library [https://github.com/rodan/ds3231](url) it looks like reading the RTC works. Now just have to see about implementing it.

Scanning...
I2C device found at address 0x3C  !
I2C device found at address 0x60  !
I2C device found at address 0x68  !
done

Date : 1/1/1900	 Hour : 0:5.29

**Reply by Dakkaron** [reply]
Nice work!

One thing just occurred to me: when only using the 4 pin edge of the DS3231, it has the same pin positions as the clock gen. So it should be possible to desolder the original headers, get some long male headers and just stack both modules on top of each other. Just squish all PCBs as close together as possible and just put a layer of insulating tape between each PCB to stop shorts.

I2C is perfect for this kind of application.

**Reply by majorpbx** [reply]
that would but the problem is the location and position of the Clock Gen board. If you did stack anything it's on the wrong side of the board. what would work better is mounting it to the other side of the clock gen board some how. Regardless, my solution with the PCB's as noted above I've already ordered. The nice thing is that you can just use a hot air rework station to de-solder and solder the same components to my board. Does require a bit of patients and experience but I think it's the easiest solution.

Once I get the code working I'm going to come up with a mod board for the Clock Gen that should be less expensive. That or I'll find a less accurate RTC to use that is more cost effective. The nice thing about RTC's is that most use the same 5 pins (SDA, SCL, VCC, GND, and Battery Connection) and the same code for most of them. That means replacements are an option and cost can be minimized.

**Reply by majorpbx** [reply]
Here is my current setup which if you install a battery and Wrap it with electrical tape you can easily store it out of the way.

![80CC944D-E1C6-44D7-9993-CFDA9C1BD232](https://user-images.githubusercontent.com/18235905/134789795-92f8b87c-89bd-4fc4-ae44-8492a05a24a5.jpeg)

**Reply by majorpbx** [reply]
Got the code to work! Now the pull requests I've submitted will allow the SD Card operations to get the newest Date/Time from the RTC and use that data for File Creation/Modification dates. I've also defaulted the RTC add-on to disabled in the options.h file.

**Reply by majorpbx** [reply]
There is a caveat to using an RTC. That is it must have a battery or time will not be right.

**Reply by Dakkaron** [reply]
For my setup I don't need the clockgen, since I only have a GBC, so I wouldn't be able to directly reproduce that setup.

I see what you mean with the problem with stacking with your board. If you use a board like this one  (https://www.amazon.de/AZDelivery-Real-Clock-DS3231-Parent/dp/B07ZQGBH14), the pins should line up fine

**Reply by Dakkaron** [reply]
Thanks for the code addition, I'll probably add an RTC to me setup some time

**Reply by majorpbx** [reply]
Cool, Just remember the pins that you need to use. The first four pins GND, VIN, SDA, SCL only if you want to just add RTC.

**Reply by majorpbx** [reply]
Running a few more tests before I do the rest of the pull requests.

**Reply by Androxilogin** [reply]
I found these in the states: **[DS3231](https://www.amazon.com/SongHe-AT24C32-Arduino-Without-Battery/dp/B082VL4GM6/ref=sr_1_1?dchild=1&keywords=azdelivery+real+time+clock+rtc+ds3231+i2c+real+time+clock&qid=1632699477&sr=8-1)**. There was some idea I had in the past that would require RTC but I forget what it was. Can you show the display? I always thought this would be a cool addition.

**Reply by majorpbx** [reply]
The code updates I'm testing only allow you to access the RTC and get an accurate Date/Time. That and it enables the SD card library to get an accurate time for timestamps. I haven't done anything with the LCD display like displaying time or anything. I'm not sure I'd go that far unless you want to be reminded it's midnight! :grin: Plus with the current option set there is really no room on the display. That is unless we start a scrolling display. That would be a good idea for a 32bit version on this project at some point though.

That board could work but you would have to do some work to hide in the case enough. Like I mentioned above my board design incorporates what you need while trying to save space. Obviously at some point I'd like to have a go at updating the main board to add functionality for all Add-on boards. It would cut down on space for sure and probably allow for other improvements.

@Androxilogin can you elaborate more on what display you are referring too?

**Reply by Androxilogin** [reply]
Ah, you answered my question which was wondering if it also displayed the time on the LCD. Now my only question is: how do you go about setting the time on this device? I ordered a few of these boards, I'm sure they'll come in handy down the line.

**Reply by majorpbx** [reply]
Ah ok :grin: glad I answered your question. I'm going to update the needed files in the source. Once I do that all anyone would need to do is wire up the RTC correctly to the Cart Reader board. Then download the new source code to your "V6.6_Portable\Arduino IDE Portable\portable\sketchbook\Cart_Reader" (for example) folder and upload the sketch to the Arduino. 

Don't forget to uncomment the options line for RTC_installed as well before you upload the sketch.

**Reply by Androxilogin** [reply]
But it grabs the time automatically once updated? I just got mine in the mail today and might end up soldering it in tonight to give it a try. Awesome work!

**Reply by majorpbx** [reply]
It does have a time but it’s not used. At present it simply uses the same time stamp across all SD save actions.

I will post here once i have submitted the final file modifications. I’ve been busy lately and i want to make sure it will work. The biggest thing is making sure it’s know what chip it will support.

I found another DS1307 data sheet but it indicates requiring 5V to operate normally. So this will need more time to iron out more support.

**Reply by majorpbx** [reply]
As for testing I received my test boards today and need to give them a go. Once i confirm it works I want to tweak the boad and provide a link to the files if anyone else wants them.

**Reply by majorpbx** [reply]
Current output of at least the n64log.txt file. It doesn't show it here to well but the data lines up tab style.

Name	: MARIOKART64
ID	: NKTE
ROM CRC1: 3e5055b6
ROM CRC2: 2e92da52
Size	: 12MB
Save	: 4K Eeprom
Version	: 1.0
Saved To: N64/ROM/MARIOKART64/857
Dumped	: 01OCT2021 03:17:32PM
CRC	: 434389c1 [Match]
Time	: 61

Here are a few pictures of the PCB's I have. Unpopulated for now, will populate one as soon as I can.
Front
![IMG_9790](https://user-images.githubusercontent.com/18235905/135675910-85827719-e06d-4a15-9773-371577117e51.jpg)

Back:
![IMG_9791](https://user-images.githubusercontent.com/18235905/135675951-ef2d0620-7e1a-4590-a178-1477877af0b1.jpg)

**Reply by majorpbx** [reply]
Posted the changes, please keep in mind this only supports DS3231. Swapping it out for DS1307 if it can actually operate at 3V at least for my board should be possible.

**Comment by Icesythe7** [comment]
Did this all work? This would be a nice addition for gbc\gba games to allow for dumping and setting of rtc related things for the mbc3 and gba games that support it.

**Reply by majorpbx** [reply]
It does though I'm not familiar enough with GBC/GBA rtc games. The same could likely be done with N64 and other carts that use rtc. I'd have to look closer at the code to see what if anything can be done. But maybe someone else has more knowledge on this and can offer suggestions. Also, I don't think I can test this on any other system than N64 as I have most carts that use rtc. I don't have any GBC/GBA titles that have rtc.

**Reply by Ancyker** [reply]
If you added networking instead of an RTC you could use NTP to get the current time as well as for saving files to network shares and such. More work, but also more features, however, I'm not sure of the memory situation on the microcontroller.

**Reply by majorpbx** [reply]
True, though that would add way more effort than a general person could commit to building one of these bad boys. As you stated memory would need a serious bump up and a way more powerful micro controller.

**Reply by Ancyker** [reply]
> As you stated memory would need a serious bump up and a way more powerful micro controller.

I'm actually working on that. :3

**Reply by majorpbx** [reply]
Awesome! Can't wait to see what you come up with.
